doc/decisions/DR016-One-to-many-result-paging.md

```diff
@@ -0,0 +1,550 @@
+# DR016 One-to-many result paging
+
+- **Status:** Proposed
+- **Impact:** Medium
+- **Driver:** @feltech
+- **Approver:** @elliotcmorris @foundrytom
+- **Contributors:** @feltech @elliotcmorris
+- **Informed:** @mattdaw
+- **Outcome:** Split batching from paging and page using a cursor object
+
+## Background
+
+The OpenAssetIO API contains functions that can return an arbitrarily
+long list of results.  In particular, the current core API has functions
+for retrieving a list of versions of a given entity, and retrieving a
+list of entities that are "related to" another entity in some way.
+
+Taking the related entities query as an example: this could be sibling
+or parent-child relationships (e.g. shot and sequence), or more loose
+relationships (e.g. the project file that was executed to generate a
+render).
+
+The current API design provides the host with a list of entity
+references that correspond to an input entity-relationship pair.
+
+There may be multiple entity-relationship pairs provided as input to
+a single API call, allowing the manager plugin to leverage batching
+optimizations. Note that, typically, batching primarily addresses
+network latency concerns, in addition to database query optimizations.
+
+In the most general case, the host cannot know the size of the list of
+related entity references that will be returned to it. This clearly
+poses a scalability issue.
+
+Database and web APIs have had to deal with this problem for decades,
+and invariably provide some mechanism of returning slices of result sets
+or "pages" that can be iterated through, lazily fetching chunks of data
+in a controlled way.
```

Comment:
[Consider] I realise we no longer (or never did?) mention stability as a problem in the Background. I wonder if we should prime the concept before we start mentioning it in pros and cons, especially as it's a large part of the reason for rejecting offset.

Reply:
Originally it was mentioned in the Background, since we enumerated the three different paging styles and made brief mention of the stability issues of offset-based solutions.

I've added a little paragraph on the features of an ideal paging solution in a6785840